src/uid2DiiNormalization.ts

```diff
@@ -0,0 +1,78 @@
+enum EmailParsingState {
+  Starting,
+  SubDomain,
+}
+
+export type NormalizedPhone = string;
+export type NormalizedEmail = string;
+
+export function isNormalizedPhone(phone: string): phone is NormalizedPhone {
+  return /^\+[0-9]{10,15}$/.test(phone);
+}
+
+export function normalizeEmail(email: string): NormalizedEmail | undefined {
+  if (email == undefined || email.length <= 0) {
+    return undefined;
+  }
+
+  let preSb = "";
+  let preSbSpecialized = "";
+  let sb = "";
+  let wsBuffer = "";
+
+  let parsingState = EmailParsingState.Starting;
+  let inExtension = false;
+
+  for (const cGiven of email) {
+    let c = cGiven;
+    if (cGiven >= "A" && cGiven < "Z") {
+      c = String.fromCharCode(c.charCodeAt(0) + 32);
+    }
+
+    switch (parsingState) {
+      case EmailParsingState.Starting:
+        if (c == " ") {
+          break;
+        }
+        if (c == "@") {
+          parsingState = EmailParsingState.SubDomain;
+        } else if (c == ".") {
+          preSb += c;
+        } else if (c == "+") {
+          preSb += c;
+          inExtension = true;
+        } else {
+          preSb += c;
+          if (!inExtension) {
+            preSbSpecialized += c;
+          }
+        }
+        break;
+      case EmailParsingState.SubDomain:
+        if (c == "@") {
+          return undefined;
+        } else if (c == " ") {
+          wsBuffer += c;
+          break;
+        } else if (wsBuffer.length > 0) {
+          sb += wsBuffer;
+          wsBuffer = "";
+        }
+        sb += c;
+    }
+  }
+
+  if (sb.length == 0) {
+    return undefined;
+  }
+
+  let domainPart = sb;
+  const GMAILDOMAIN = "gmail.com";
```

Comment:
nit: maybe GMAIL_DOMAIN to align with other const?

Reply:
Sorry, this entire function is more lifted code 😅 This time from https://github.com/UnifiedID2/uid2-snowflake/blob/99c6346ed4c5b4fceadfec4c3e2407094f6804c7/snowflake/functions.sql#L16-L88.

Reply:
oooh haha gotcha, no worries